File: credentials-verification-android/app/src/main/java/io/iohk/cvp/viewmodel/CredentialsViewModel.java
package io.iohk.cvp.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import io.iohk.cvp.grpc.AsyncTaskResult;
import io.iohk.cvp.grpc.GetConnectionsInfoRunnable;
import io.iohk.cvp.grpc.GetMessagesRunnable;
import io.iohk.cvp.grpc.GrpcTask;
import io.iohk.prism.protos.ConnectionInfo;
import io.iohk.prism.protos.ReceivedMessage;

public class CredentialsViewModel extends NewConnectionsViewModel {

    private final MutableLiveData<AsyncTaskResult<List<ReceivedMessage>>> messages = new MutableLiveData<>(
            new AsyncTaskResult<>());
    private MutableLiveData<AsyncTaskResult<List<ConnectionInfo>>> connections = new MutableLiveData<>();

    @Inject
    public CredentialsViewModel() {
    }

    public void clearMessages() {
        this.messages.setValue(
                new AsyncTaskResult<>(new ArrayList<>()));
    }

    public MutableLiveData<AsyncTaskResult<List<ReceivedMessage>>> getMessages(Set<String> userIds) {
        GrpcTask task = new GrpcTask<>(new GetMessagesRunnable(messages), context);
        task.execute(userIds);
        runningTasks.add(task);
        return messages;
    }

    public LiveData<AsyncTaskResult<List<ConnectionInfo>>> getConnections(Set<String> userIds) {
        GrpcTask task = new GrpcTask<>(new GetConnectionsInfoRunnable(connections), context);
        task.execute(userIds);
        runningTasks.add(task);
        return connections;
    }

}

File: credentials-verification-android/app/src/main/java/io/iohk/cvp/viewmodel/CredentialsViewModel.kt
package io.iohk.cvp.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.crashlytics.FirebaseCrashlytics
import io.iohk.cvp.data.DataManager
import io.iohk.cvp.data.local.db.model.Credential
import io.iohk.cvp.grpc.AsyncTaskResult
import io.iohk.cvp.views.utils.SingleLiveEvent
import kotlinx.coroutines.*
import java.lang.Exception
import javax.inject.Inject

class CredentialsViewModel @Inject constructor(val dataManager: DataManager) : NewConnectionsViewModel(dataManager) {

    private val _credentials = MutableLiveData(AsyncTaskResult<List<Credential>>())

    private val _credentialDeletedLiveData = MutableLiveData<Boolean>(false)

    private val _showErrorMessageLiveData = SingleLiveEvent<Boolean>()

    fun getNewCredentials() {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val credentialsList = dataManager.getAllNewCredentials()
                _credentials.postValue(AsyncTaskResult(credentialsList))
            } catch (ex: Exception) {
                _credentials.postValue(AsyncTaskResult(ex))
            }
        }
    }

    fun getAllCredentials() {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val credentialsList = dataManager.getAllCredentials()
                _credentials.postValue(AsyncTaskResult(credentialsList))
            } catch (ex: Exception) {
                _credentials.postValue(AsyncTaskResult(ex))
            }
        }
    }

    fun getCredentialLiveData(): LiveData<AsyncTaskResult<List<Credential>>> {
        return _credentials
    }

    fun setCredentialViewed(credential: Credential) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                credential.viewed = true
                dataManager.updateCredential(credential)
            } catch (ex: Exception) {
                FirebaseCrashlytics.getInstance().recordException(ex)
            }
        }
    }

    fun getDeleteCredentialLiveData(): LiveData<Boolean> {
        return _credentialDeletedLiveData;
    }

    fun getShowErrorMessageLiveData(): SingleLiveEvent<Boolean> {
        return _showErrorMessageLiveData;
    }

    fun deleteCredential(credentialId: String) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                val credential = dataManager.getCredentialByCredentialId(credentialId)
                dataManager.deleteCredential(credential)
                _credentialDeletedLiveData.postValue(true)
            } catch (ex: Exception) {
                FirebaseCrashlytics.getInstance().recordException(ex)
                _showErrorMessageLiveData.postValue(true)
            }
        }
    }
}
